packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -688,154 +861,585 @@ class _SortData {
+      if (order != null) {
+        ordered.add(_OrderedFocusInfo(node: node, order: order));
+      } else {
+        unordered.add(node);
       }
     }
-    final List<FocusNode> sortedNodes = _sortByGeometry(nearestScope).toList();
-    if (forward && focusedChild == sortedNodes.last) {
-      _focusAndEnsureVisible(sortedNodes.first, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtEnd);
-      return true;
-    }
-    if (!forward && focusedChild == sortedNodes.first) {
-      _focusAndEnsureVisible(sortedNodes.last, alignmentPolicy: ScrollPositionAlignmentPolicy.keepVisibleAtStart);
-      return true;
-    }
+    mergeSort<_OrderedFocusInfo>(ordered, compare:(_OrderedFocusInfo a, _OrderedFocusInfo b) {
+      assert(a.order.runtimeType == b.order.runtimeType,
+        'When sorting nodes for determining focus order, the order (${a.order}) of '
+        "node ${a.node}, isn't the same type as the order (${b.order}) of ${b.node}. "
+        "Incompatible order types can't be compared.  Use a FocusTraversalGroup to group "
+        'similar orders together.');
+      return a.order.compareTo(b.order);
+    });
+    return ordered.map<FocusNode>((_OrderedFocusInfo info) => info.node).followedBy(unordered);
+  }
+}
 
-    final Iterable<FocusNode> maybeFlipped = forward ? sortedNodes : sortedNodes.reversed;
-    FocusNode previousNode;
-    for (final FocusNode node in maybeFlipped) {
-      if (previousNode == focusedChild) {
-        _focusAndEnsureVisible(
-          node,
-          alignmentPolicy: forward
-            ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-            : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
-        );
-        return true;
+/// An inherited widget that describes the order in which its child subtree
+/// should be traversed.
+///
+/// {@macro flutter.widgets.focusorder.comparable}
+///
+/// The order for a widget is determined by the [FocusOrder] returned by
+/// [FocusTraversalOrder.of] for a particular context.
+class FocusTraversalOrder extends InheritedWidget {
+  /// A const constructor so that subclasses can be const.
+  const FocusTraversalOrder({Key key, this.order, Widget child}) : super(key: key, child: child);
+
+  /// The order for the widget descendants of this [FocusTraversalOrder].
+  final FocusOrder order;
+
+  /// Finds the nearest TraversalOrder widget.
+  ///
+  /// It does not create a rebuild dependency because changing the traversal
+  /// order doesn't change the widget tree, so nothing needs to be rebuilt as a
+  /// result of an order change.
+  static FocusOrder of(BuildContext context, {bool nullOk = false}) {
+    assert(context != null);
+    assert(nullOk != null);
+    final FocusTraversalOrder marker = context.getElementForInheritedWidgetOfExactType<FocusTraversalOrder>()?.widget as FocusTraversalOrder;
+    final FocusOrder order = marker?.order;
+    if (order == null) {
```

Comment:
you could fold both `if`s into one and, remove the `return null` and then just return `order` after the if, which may just be null.

Reply:
Thanks, consolidated.